File: python/test/mod_test_shape_truthful_clustering_csv.py
"""
Python test suite for tox_stc_csv (serialize_stc_points_as_csv,
serialize_stc_ensemble_overlap_as_csv, serialize_stc_super_ensembles_as_tsv), mirroring
test/mod_test_shape_truthful_clustering_csv.F90
"""

import numpy as np
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
from tensor_omics import (
    serialize_stc_points_as_csv,
    serialize_stc_ensemble_overlap_as_csv,
    serialize_stc_super_ensembles_as_tsv,
)
from test_helpers import run_all_tests


def _fixture():
    """N=4, 2 ensembles: {1,2,3} (seed=1) and {2,3,4} (seed=4), overlapping on {2,3} --
    Overlap Coefficient 2/3 -- merged into one super-ensemble. Point 1 is also flagged
    low-confidence. Matches mod_test_shape_truthful_clustering_json.py's own fixture."""
    seed_selection_mask = np.array([True, False, False, True], dtype=np.bool_)
    ensemble_masks = np.zeros((4, 2), dtype=np.bool_, order='F')
    ensemble_masks[0:3, 0] = True
    ensemble_masks[1:4, 1] = True
    ensemble_low_confidence_masks = np.zeros((4, 2), dtype=np.bool_, order='F')
    ensemble_low_confidence_masks[0, 0] = True
    super_ensembles = np.asfortranarray([[1, 0], [2, 0]], dtype=np.int32)
    return seed_selection_mask, ensemble_masks, ensemble_low_confidence_masks, super_ensembles


def test_points_csv_membership():
    seed_selection_mask, ensemble_masks, ensemble_low_confidence_masks, super_ensembles = _fixture()

    filename = "test_stc_points_py.csv"
    serialize_stc_points_as_csv(filename, 1, seed_selection_mask, ensemble_masks,
                                ensemble_low_confidence_masks, super_ensembles)

    with open(filename, "r") as f:
        lines = [line.rstrip("\n") for line in f.readlines()]
    os.remove(filename)

    assert lines[0] == "row,ensembles,super_ensembles,low_confidence_ensembles,seed_of"
    assert lines[1] == '1,"1","1","1","1"'
    assert lines[2] == '2,"1,2","1","",""'
    assert lines[3] == '3,"1,2","1","",""'
    assert lines[4] == '4,"2","1","","2"'


def test_points_csv_zero_ensembles():
    seed_selection_mask = np.array([False, False], dtype=np.bool_)
    ensemble_masks = np.zeros((2, 0), dtype=np.bool_, order='F')
    ensemble_low_confidence_masks = np.zeros((2, 0), dtype=np.bool_, order='F')
    super_ensembles = np.zeros((2, 0), dtype=np.int32, order='F')

    filename = "test_stc_points_zero_py.csv"
    serialize_stc_points_as_csv(filename, 0, seed_selection_mask, ensemble_masks,
                                ensemble_low_confidence_masks, super_ensembles)

    with open(filename, "r") as f:
        lines = [line.rstrip("\n") for line in f.readlines()]
    os.remove(filename)

    assert lines[1] == '1,"","","",""'


def test_overlap_csv_pairwise_coefficient():
    _, ensemble_masks, _, _ = _fixture()

    filename = "test_stc_overlap_py.csv"
    serialize_stc_ensemble_overlap_as_csv(filename, ensemble_masks)

    with open(filename, "r") as f:
        lines = [line.rstrip("\n") for line in f.readlines()]
    os.remove(filename)

    assert lines[0] == "ensemble_a,ensemble_b,overlap_coefficient"
    assert lines[1] == "1,2,6.6666666666666663E-001"


def test_overlap_csv_no_intersecting_pairs():
    ensemble_masks = np.zeros((4, 2), dtype=np.bool_, order='F')
    ensemble_masks[0:2, 0] = True
    ensemble_masks[2:4, 1] = True

    filename = "test_stc_overlap_disjoint_py.csv"
    serialize_stc_ensemble_overlap_as_csv(filename, ensemble_masks)

    with open(filename, "r") as f:
        content = f.read()
    os.remove(filename)

    assert content.strip() == "ensemble_a,ensemble_b,overlap_coefficient"


def test_super_ensembles_tsv_gene_family_format():
    super_ensembles = np.asfortranarray([[1], [2], [0]], dtype=np.int32)

    filename = "test_stc_super_ensembles_py.tsv"
    serialize_stc_super_ensembles_as_tsv(filename, super_ensembles)

    with open(filename, "r") as f:
        lines = [line.rstrip("\n") for line in f.readlines()]
    os.remove(filename)

    assert lines[0] == "1\t1,2"


def test_super_ensembles_tsv_empty_when_zero_groups():
    super_ensembles = np.zeros((2, 0), dtype=np.int32, order='F')

    filename = "test_stc_super_ensembles_empty_py.tsv"
    serialize_stc_super_ensembles_as_tsv(filename, super_ensembles)

    with open(filename, "r") as f:
        content = f.read()
    os.remove(filename)

    assert content.strip() == ""


if __name__ == "__main__":
    run_all_tests(globals().values())

File: r/test/mod_test_shape_truthful_clustering_csv.R
source("r/load_tensor_omics.R")
source("r/test_helpers.R")

# N=4, 2 ensembles: {1,2,3} (seed=1) and {2,3,4} (seed=4), overlapping on {2,3} -- Overlap
# Coefficient 2/3 -- merged into one super-ensemble. Point 1 is also flagged low-confidence.
# Matches mod_test_shape_truthful_clustering_json.R's own fixture.
fixture <- function() {
  seed_selection_mask <- c(TRUE, FALSE, FALSE, TRUE)
  ensemble_masks <- matrix(FALSE, nrow = 4, ncol = 2)
  ensemble_masks[1:3, 1] <- TRUE
  ensemble_masks[2:4, 2] <- TRUE
  ensemble_low_confidence_masks <- matrix(FALSE, nrow = 4, ncol = 2)
  ensemble_low_confidence_masks[1, 1] <- TRUE
  super_ensembles <- matrix(c(1, 2, 0, 0), nrow = 2, ncol = 2)
  list(seed_selection_mask = seed_selection_mask, ensemble_masks = ensemble_masks,
       ensemble_low_confidence_masks = ensemble_low_confidence_masks, super_ensembles = super_ensembles)
}

test_points_csv_membership <- function() {
  fx <- fixture()
  filename <- "test_stc_points_r.csv"
  serialize_stc_points_as_csv(filename, 1L, fx$seed_selection_mask, fx$ensemble_masks,
                               fx$ensemble_low_confidence_masks, fx$super_ensembles)

  lines <- readLines(filename, warn = FALSE)
  file.remove(filename)

  assert_true(lines[1] == "row,ensembles,super_ensembles,low_confidence_ensembles,seed_of")
  assert_true(lines[2] == '1,"1","1","1","1"')
  assert_true(lines[3] == '2,"1,2","1","",""')
  assert_true(lines[4] == '3,"1,2","1","",""')
  assert_true(lines[5] == '4,"2","1","","2"')
}

test_points_csv_zero_ensembles <- function() {
  seed_selection_mask <- c(FALSE, FALSE)
  ensemble_masks <- matrix(FALSE, nrow = 2, ncol = 0)
  ensemble_low_confidence_masks <- matrix(FALSE, nrow = 2, ncol = 0)
  super_ensembles <- matrix(0L, nrow = 2, ncol = 0)

  filename <- "test_stc_points_zero_r.csv"
  serialize_stc_points_as_csv(filename, 0L, seed_selection_mask, ensemble_masks,
                               ensemble_low_confidence_masks, super_ensembles)

  lines <- readLines(filename, warn = FALSE)
  file.remove(filename)

  assert_true(lines[2] == '1,"","","",""')
}

test_overlap_csv_pairwise_coefficient <- function() {
  fx <- fixture()
  filename <- "test_stc_overlap_r.csv"
  serialize_stc_ensemble_overlap_as_csv(filename, fx$ensemble_masks)

  lines <- readLines(filename, warn = FALSE)
  file.remove(filename)

  assert_true(lines[1] == "ensemble_a,ensemble_b,overlap_coefficient")
  assert_true(lines[2] == "1,2,6.6666666666666663E-001")
}

test_overlap_csv_no_intersecting_pairs <- function() {
  ensemble_masks <- matrix(FALSE, nrow = 4, ncol = 2)
  ensemble_masks[1:2, 1] <- TRUE
  ensemble_masks[3:4, 2] <- TRUE

  filename <- "test_stc_overlap_disjoint_r.csv"
  serialize_stc_ensemble_overlap_as_csv(filename, ensemble_masks)

  lines <- readLines(filename, warn = FALSE)
  file.remove(filename)

  non_empty <- lines[nzchar(lines)]
  assert_true(length(non_empty) == 1)
  assert_true(non_empty[1] == "ensemble_a,ensemble_b,overlap_coefficient")
}

test_super_ensembles_tsv_gene_family_format <- function() {
  super_ensembles <- matrix(c(1, 2, 0), nrow = 3, ncol = 1)

  filename <- "test_stc_super_ensembles_r.tsv"
  serialize_stc_super_ensembles_as_tsv(filename, super_ensembles)

  lines <- readLines(filename, warn = FALSE)
  file.remove(filename)

  assert_true(lines[1] == "1\t1,2")
}

test_super_ensembles_tsv_empty_when_zero_groups <- function() {
  super_ensembles <- matrix(0L, nrow = 2, ncol = 0)

  filename <- "test_stc_super_ensembles_empty_r.tsv"
  serialize_stc_super_ensembles_as_tsv(filename, super_ensembles)

  lines <- readLines(filename, warn = FALSE)
  file.remove(filename)

  non_empty <- lines[nzchar(lines)]
  assert_true(length(non_empty) == 0)
}

run_all_tests()
